roles/translator/src/template_receiver/mod.rs

```diff
@@ -0,0 +1,142 @@
+        solution_receiver: Receiver<SubmitSolution<'static>>,
+    ) {
+        let stream = async_std::net::TcpStream::connect(address).await.unwrap();
+
+        let (mut receiver, mut sender): (Receiver<EitherFrame>, Sender<EitherFrame>) =
+            PlainConnection::new(stream, 10).await;
+
+        SetupConnectionHandler::setup(&mut receiver, &mut sender, address)
+            .await
+            .unwrap();
+
+        let (coinbase_output_max_additional_size, token) =
+            receive_coinbase_output_max_additional_size
+                .recv()
+                .await
+                .unwrap();
+
+        let self_mutex = Arc::new(Mutex::new(Self {
+            receiver: receiver.clone(),
+            sender: sender.clone(),
+            send_new_tp_to_negotiator,
+            send_new_ph_to_negotiator,
+        }));
+
+        let sv2_frame: StdFrame = PoolMessages::TemplateDistribution(
+            roles_logic_sv2::parsers::TemplateDistribution::CoinbaseOutputDataSize(
+                coinbase_output_max_additional_size,
+            ),
+        )
+        .try_into()
+        .unwrap();
+        Self::send(self_mutex.clone(), sv2_frame).await;
+        let cloned = self_mutex.clone();
+
+        tokio::task::spawn(Self::on_new_solution(cloned, solution_receiver));
+        Self::start_templates(self_mutex, token);
+    }
+
+    pub async fn send(self_: Arc<Mutex<Self>>, sv2_frame: StdFrame) {
+        let either_frame = sv2_frame.into();
+        let sender_to_tp = self_.safe_lock(|self_| self_.sender.clone()).unwrap();
+        match sender_to_tp.send(either_frame).await {
+            Ok(_) => (),
+            Err(e) => panic!("{:?}", e),
+        }
+    }
+
+    pub fn start_templates(self_mutex: Arc<Mutex<Self>>, token: u64) {
+        tokio::task::spawn(async move {
+            // Send CoinbaseOutputDataSize size to TP
+            loop {
+                // Receive Templates and SetPrevHash from TP to send to JN
+                let receiver = self_mutex
+                    .clone()
+                    .safe_lock(|s| s.receiver.clone())
+                    .unwrap();
+                let mut frame: StdFrame = receiver.recv().await.unwrap().try_into().unwrap();
+                let message_type = frame.get_header().unwrap().msg_type();
+                let payload = frame.payload();
+
+                let next_message_to_send =
+                    ParseServerTemplateDistributionMessages::handle_message_template_distribution(
+                        self_mutex.clone(),
+                        message_type,
+                        payload,
+                    );
+                match next_message_to_send {
+                    Ok(SendTo::None(m)) => match m {
+                        Some(TemplateDistribution::NewTemplate(m)) => {
+                            crate::upstream_sv2::upstream::IS_NEW_TEMPLATE_HANDLED
+                                .store(false, std::sync::atomic::Ordering::SeqCst);
```

Comment:
Does this race condition concern you? I could see a case where NewTemplates come it pretty rapidly around when a block is found which could delay the SNPH from going out which would leave miners mining on old jobs for longer increasing stale shares. If a new template comes in and resets this to false right after the previous template was handled it will continually delay the SNPH from being processed until there is a lull in NewTemplates.

Reply:
When we have a new block the TP send (NT, SNPH) with NT future. NT is sent first and SNPH after but it could happen that the handler of SNPH is called before than the one of NT cause they are handled in different tasks. To make sure that this can not happen we use `IS_NEW_TEMPLATE_HANDLED`. Calling before the handler of SNPH is an issue cause SNPH look for a NT with an id of the one that is not already handled.

The above race condition can happen and will lead to a bug the latter NT will be discarded when we handle SNPH so if the next SNPH point to the discarded  NT we have an issue.

Reply:
maybe we can open an issue? Not sure if make sense to find a better solution now as it is very unlikley

Reply:
yeah - opening an issue works for me.